多言語化対応（英語、日本語）

damisumida opened the pull request:
「Railsのi18nを理解する」の課題を作成しました。
ご確認をよろしくお願いいたします。

damisumida committed:
多言語化対応（英語、日本語）

maedana commented on Gemfile:
```diff
@@ -59,3 +59,6 @@ gem 'net-pop'
 
 gem 'carrierwave'
+
+gem 'i18n_generators'
```
開発時だけ必要なgemなのでproductionには含まないようにしたいですねー。

READMEはご覧になりましたか?
https://github.com/amatsuda/i18n_generators#installation
でも言及されてますので確認して適切に対応してみましょう

maedana commented on app/controllers/books_controller.rb:
```diff
@@ -1,7 +1,7 @@
 
 class BooksController < ApplicationController
-  before_action :set_book, only: %i[show edit update destroy]
+  before_action :set_book, :set_locale, only: %i[show edit update destroy]
```
スクショを拝見すると
> config.i18n.default_locale

を切り替えることで言語切り替えする方針だと思いました。となるとこれ何のためにやっているのでしょうね? どういうものか理解して追加してますか?

damisumida replied:
この処理については、URLパラメータのlocaleの値から翻訳を参照する処理の認識なのですが、あっておりますでしょうか？

このため、提出物についてはlocaleをURLより指定する方針をとるべきでした・・・
また、localeをURLより指定し動作確認したところ正しく動作していなかったためこちらは提出物も含め修正いたします。

maedana replied:
その理解であってます。修正方針についてもOKです。

maedana commented on app/views/books/index.html.erb:
```diff
@@ -1,14 +1,14 @@
-      <th>Memo</th>
-      <th>Author</th>
-      <th>Picture</th>
+      <th><%= Book.human_attribute_name'title' %></th>
```
動いているのだと思いますが、括弧を省略するなら`'title'`の前空白空けたいですね。ほかも同様です。

maedana commented on app/views/books/index.html.erb:
```diff
@@ -20,14 +20,14 @@
 <br>
 
-<%= link_to 'New Book', new_book_path %>
+<%= link_to t('links.newbook'), new_book_path %>
```
`newbook`という単語は存在しないので存在する単語をスネークケースで表現しましょうー

maedana commented on app/views/layouts/application.html.erb:
```diff
@@ -1,7 +1,7 @@
 <html>
   <head>
-    <title>BooksApp</title>
+    <title><%= t '.booksapp' %></title>
```
同上です。`booksapp`という単語は存在しないかなと

maedana commented on Gemfile:
```diff
@@ -59,3 +59,6 @@ gem 'net-pop'
 gem 'carrierwave'
+
+gem 'i18n_generators'
+gem 'rails-i18n'
```
念の為確認ですが、このgemがどういうものか理解した上で追加してますか?

damisumida replied:
それぞれ以下の認識なのですがあっておりますでしょうか？
``` i18n_generators ```については、modelとcontrolerを基に翻訳ファイル（hoge.yml)を作成するツールの認識です。

``` rails-i18n ```については以下を行う認識です。
- 翻訳ファイルを基に、viewの表示を翻訳する
- localeの取得/設定

maedana replied:
説明不足ですみません。書いたコメントは直近の1行についてのみ、つまり`rails-i18n`についてのみでした。その上でご認識は違います。Rails本体のみで国際化は動作します。

Railsガイドにも言及がありますね。
https://railsguides.jp/i18n.html

> RubyのI18nフレームワークでは、Railsアプリケーションの国際化/ローカライズに必要な手段がすべて提供されています。もちろん、さまざまなプラグインや拡張機能を導入してそれ以外の機能を追加しても構いません。詳しくは[rails-i18n gem](https://github.com/svenfuchs/rails-i18n)を参照してください。

この部分もですね
> 現時点では、I18nバックエンドが（少なくとも海賊語ロケールで）本当に正常に動作するには、日付・時刻フォーマットを追加する必要があるでしょう。もちろん、誰か親切な人が Railsのデフォルト文字列をすべて翻訳してくれた結果がネット上のどこかに既にあるかもしれません。ロケールファイルのアーカイブがあるかどうかについては、[GitHubのrails-i18nリポジトリ](https://github.com/svenfuchs/rails-i18n/tree/master/rails/locale)を探してみるとよいでしょう。運よくそうしたファイルが見つかれば、config/locales/ディレクトリに置くだけで即座に使えるようになります。

また、
https://github.com/svenfuchs/rails-i18n
のREADMEを読んでみましょうー

その上で必要性を判断出来るといいのですが、いかがでしょうか?

damisumida replied:
rails-i18nにつて完全に勘違いしていました・・・。リンクありがとうございます。

見直した結果、rails-i18nは以下を行うことができるものだと認識しました。

- デフォルトの翻訳ファイルを取得できる
- ロケールごとの複数形ルールを設定する

今回はデフォルトの翻訳ファイルを使用する必要がなく、
ロケールごとに複数形ルールを設定する必要はないため、不要かと思っています。

上記、認識あっておりますでしょうか？

maedana replied:
今回は事実上影響ないのでいいのですけど、厳密には
https://github.com/svenfuchs/rails-i18n/blob/master/rails/locale/ja.yml
がないとRails本体で必要な国際化がされない、ってことになりますね。僕が先日コメントした`現時点では、I18nバックエンドが（少なくとも海賊語ロケールで）本当に正常に動作するに...`のくだりにあることです。今回の課題においてはそのへんの機能おそらく未使用で影響受けませんのでなくてもOKとします。

damisumida committed:
- i18n_generatorsをdevelopmentグループに移動
- switch_localeメソッドをApplicationControllerに移動
- view内空白を追加
- 翻訳前の文字列を修正
- en.yml内のスペルミスを修正

damisumida committed:
- locale設定箇所を画面遷移時も保持できるよう修正

damisumida commented:
ご確認ありがとうございます。
ご指摘いただいた箇所を修正いたしましたので、再度ご確認のほどよろしくお願いいたします。
